src/System.Windows.Extensions/src/System/Security/Cryptography/X509Certificates/X509Certificate2UI.cs

```diff
@@ -0,0 +1,150 @@
+                throw new ArgumentNullException(nameof(certificates));
+            if (selectionFlag < X509SelectionFlag.SingleSelection || selectionFlag > X509SelectionFlag.MultiSelection)
+                throw new ArgumentException(SR.Format(SR.Enum_InvalidValue, nameof(selectionFlag)));
+
+            using (SafeCertStoreHandle safeSourceStoreHandle = X509Utils.ExportToMemoryStore(certificates))
+            using (SafeCertStoreHandle safeTargetStoreHandle = SelectFromStore(safeSourceStoreHandle, title, message, selectionFlag, hwndParent))
+            {
+                return X509Utils.GetCertificates(safeTargetStoreHandle);
+            }
+        }
+
+        private static unsafe SafeCertStoreHandle SelectFromStore(SafeCertStoreHandle safeSourceStoreHandle, string title, string message, X509SelectionFlag selectionFlags, IntPtr hwndParent)
+        {
+            int dwErrorCode = ERROR_SUCCESS;
+
+            SafeCertStoreHandle safeCertStoreHandle = Interop.Crypt32.CertOpenStore((IntPtr)Interop.Crypt32.CERT_STORE_PROV_MEMORY,
+                                                                         Interop.Crypt32.X509_ASN_ENCODING | Interop.Crypt32.PKCS_7_ASN_ENCODING,
+                                                                         IntPtr.Zero,
+                                                                         0,
+                                                                         null);
+
+            if (safeCertStoreHandle == null || safeCertStoreHandle.IsInvalid)
+                throw new CryptographicException(Marshal.GetLastWin32Error());
+
+            Interop.CryptUI.CRYPTUI_SELECTCERTIFICATE_STRUCTW csc = new Interop.CryptUI.CRYPTUI_SELECTCERTIFICATE_STRUCTW();
+            // Older versions of CRYPTUI do not check the size correctly,
+            // so always force it to the oldest version of the structure.
+            csc.dwSize = (uint)Marshal.OffsetOf(typeof(Interop.CryptUI.CRYPTUI_SELECTCERTIFICATE_STRUCTW), "hSelectedCertStore");
+            csc.hwndParent = hwndParent;
+            csc.dwFlags = (uint)selectionFlags;
+            csc.szTitle = title;
+            csc.dwDontUseColumn = 0;
+            csc.szDisplayString = message;
+            csc.pFilterCallback = IntPtr.Zero;
+            csc.pDisplayCallback = IntPtr.Zero;
+            csc.pvCallbackData = IntPtr.Zero;
+            csc.cDisplayStores = 1;
+            IntPtr hSourceCertStore = safeSourceStoreHandle.DangerousGetHandle();
+            csc.rghDisplayStores = new IntPtr(&hSourceCertStore);
+            csc.cStores = 0;
+            csc.rghStores = IntPtr.Zero;
+            csc.cPropSheetPages = 0;
+            csc.rgPropSheetPages = IntPtr.Zero;
+            csc.hSelectedCertStore = safeCertStoreHandle.DangerousGetHandle();
+
+            SafeCertContextHandle safeCertContextHandle = Interop.CryptUI.CryptUIDlgSelectCertificateW(csc);
+
+            if (safeCertContextHandle != null && !safeCertContextHandle.IsInvalid)
```

Comment:
In what situation would this be null?

Reply:
I also thought these checks are redundant. But these are present on desktop, so i decided to keep them until we have addded more tests